File: 3.Longest Substring Without Repeating Characters/longest_substring_without_repeating_characters_2.go
package main

import "fmt"

func lengthOfLongestSubstring(s string) int {
	start_cursor, max_length := -1, 0
	// use map keys cannot be same
	charsMap := make(map[byte]int)

	for i := 0; i < len(s); i++ {
		if _, ok := charsMap[s[i]]; ok && charsMap[s[i]] > start_cursor {
			// print the string without repeating chars for now
			// fmt.Println(charsMap)
			// let start_cursor move to char ch last one position before this loop
			start_cursor = charsMap[s[i]]
			charsMap[s[i]] = i
		} else {
			charsMap[s[i]] = i
			max_length = max(max_length, i-start_cursor)
		}
	}

	return max_length

}

func max(x int, y int) int {
	if x >= y {
		return x
	}
	return y
}

func main() {
	s := "dfasdfwefwfe"
	rest := lengthOfLongestSubstring(s)
	fmt.Println(rest)
}

File: 3. Longest Substring Without Repeating Characters/longest_substring_without_repeating_characters_2.py


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        start_cursor, max_length, char_dict = -1, 0, {}

        for index, ch in enumerate(s):
            if ch in char_dict and char_dict[ch] > start_cursor:
                # print the string without repeating chars for now
                # print(char_dict)
                # let start_cursor move to char ch last one position before this loop
                start_cursor = char_dict[ch]
                char_dict[ch] = index
            else:
                char_dict[ch] = index
                max_length = max(max_length, index-start_cursor)
        return max_length


def max(x: int, y: int) -> int:
    if x > y:
        return x
    return y


def main():
    s = Solution()
    print(s.lengthOfLongestSubstring("asdfasdfwefad"))


if __name__ == "__main__":
    main()
